File: parse/src/parsers/book_parser.py
"""Book page HTML parser for Goodreads.

Extracts detailed book metadata from individual Goodreads book pages.
Used for extended metadata (User Stories 2 & 3). MVP uses library_parser primarily.
Per Constitution Principle I: Data-First with extensible parsing.
"""

from bs4 import BeautifulSoup
from typing import Any
import json
import re

from src.validators import sanitize_text, validate_isbn, validate_publication_year


def parse_book_page(html: str) -> dict[str, Any]:
    """Parse a Goodreads book page HTML for detailed metadata.

    This is used for extracting extended metadata (ISBN, genres, publication details)
    that may not be available on the library list page.

    Args:
        html: Raw HTML content from Goodreads book page

    Returns:
        Dictionary containing book metadata:
        - isbn: ISBN number
        - isbn13: ISBN-13 format
        - publication_year: Year published
        - publisher: Publisher name
        - page_count: Number of pages
        - genres: List of genre tags
        - language: Book language
        - setting: Book setting/location
        - literary_awards: List of literary awards (dicts with name, category, year)
        - average_rating: Goodreads community rating
        - ratings_count: Number of ratings
        - cover_image_url: URL to book cover image

    Note:
        For MVP (User Story 1), this is primarily a stub.
        Full implementation in User Stories 2 & 3 for extended metadata.
    """
    soup = BeautifulSoup(html, 'lxml')

    book_data = {}

    # First try to extract from Next.js __NEXT_DATA__ (most complete source, includes publisher)
    next_data = extract_from_next_data(soup)
    if next_data:
        book_data.update(next_data)

    # Then try JSON-LD schema as fallback
    schema_data = extract_from_schema(soup)
    if schema_data:
        # Only update fields not already set from __NEXT_DATA__
        for key, value in schema_data.items():
            if value and not book_data.get(key):
                book_data[key] = value

    # Extract ISBN (User Story 2) - fallback if not in schema
    if not book_data.get('isbn') and not book_data.get('isbn13'):
        isbn_data = extract_isbn(soup)
        book_data.update(isbn_data)

    # Extract publication details (User Story 2)
    pub_data = extract_publication_details(soup)
    # Only update fields that aren't already set from schema
    for key, value in pub_data.items():
        if value and not book_data.get(key):
            book_data[key] = value

    # Extract genres (User Story 2)
    if not book_data.get('genres'):
        genres = extract_genres(soup)
        book_data['genres'] = genres

    # Extract ratings (available on library page too, but more detailed here)
    if not book_data.get('average_rating'):
        rating_data = extract_rating_details(soup)
        book_data.update(rating_data)

    # Extract cover image
    if not book_data.get('cover_image_url'):
        cover_url = extract_cover_image(soup)
        book_data['cover_image_url'] = cover_url

    return book_data


def extract_from_next_data(soup: BeautifulSoup) -> dict[str, Any]:
    """Extract book metadata from Next.js __NEXT_DATA__ Apollo state.

    This contains detailed book information including publisher, setting, and awards.

    Args:
        soup: BeautifulSoup parsed HTML

    Returns:
        Dictionary with book metadata from Apollo state
    """
    next_data = {}

    # Find __NEXT_DATA__ script tag
    script_tag = soup.find('script', id='__NEXT_DATA__')
    if script_tag and script_tag.string:
        try:
            data = json.loads(script_tag.string)
            apollo_state = data.get('props', {}).get('pageProps', {}).get('apolloState', {})

            # Find the Book object (key starts with "Book:")
            for key, value in apollo_state.items():
                if key.startswith('Book:') and isinstance(value, dict):
                    # Extract details which contains publisher, ISBN, etc.
                    details = value.get('details')
                    if details and isinstance(details, dict):
                        # Publisher
                        if 'publisher' in details:
                            next_data['publisher'] = sanitize_text(details['publisher'], max_length=200)

                        # ISBN and ISBN-13
                        if 'isbn' in details:
                            isbn_str = str(details['isbn'])
                            validated_isbn = validate_isbn(isbn_str)
                            if validated_isbn and len(validated_isbn) == 10:
                                next_data['isbn'] = validated_isbn

                        if 'isbn13' in details:
                            isbn13_str = str(details['isbn13'])
                            validated_isbn13 = validate_isbn(isbn13_str)
                            if validated_isbn13 and len(validated_isbn13) == 13:
                                next_data['isbn13'] = validated_isbn13

                        # Page count
                        if 'numPages' in details:
                            try:
                                next_data['page_count'] = int(details['numPages'])
                            except (ValueError, TypeError):
                                pass

                        # Language
                        if 'language' in details and isinstance(details['language'], dict):
                            lang_name = details['language'].get('name')
                            if lang_name:
                                next_data['language'] = sanitize_text(lang_name)

                        # Publication year from publicationTime (milliseconds timestamp)
                        if 'publicationTime' in details:
                            try:
                                timestamp_ms = int(details['publicationTime'])
                                from datetime import datetime
                                pub_date = datetime.fromtimestamp(timestamp_ms / 1000)
                                next_data['publication_year'] = validate_publication_year(pub_date.year)
                            except (ValueError, TypeError, OSError):
                                pass

                    break  # Only process first Book object

            # Find the Work object (key starts with "Work:") for setting and awards
            for key, value in apollo_state.items():
                if key.startswith('Work:') and isinstance(value, dict):
                    details = value.get('details')
                    if details and isinstance(details, dict):
                        # Setting - extract from places array
                        if 'places' in details and isinstance(details['places'], list) and len(details['places']) > 0:
                            first_place = details['places'][0]
                            if isinstance(first_place, dict) and 'name' in first_place:
                                setting_name = first_place['name']
                                if setting_name:
                                    next_data['setting'] = sanitize_text(setting_name, max_length=200)

                        # Literary Awards - extract from awardsWon array
                        if 'awardsWon' in details and isinstance(details['awardsWon'], list):
                            awards = []
                            for award_data in details['awardsWon']:
                                if isinstance(award_data, dict):
                                    award_name = award_data.get('name')
                                    if award_name:
                                        # Extract category (may be in 'category' field)
                                        category = award_data.get('category')

                                        # Extract year from awardedAt timestamp (milliseconds)
                                        year = None
                                        if 'awardedAt' in award_data:
                                            try:
                                                timestamp_ms = int(award_data['awardedAt'])
                                                from datetime import datetime
                                                award_date = datetime.fromtimestamp(timestamp_ms / 1000)
                                                year = award_date.year
                                            except (ValueError, TypeError, OSError):
                                                pass

                                        awards.append({
                                            'name': sanitize_text(award_name, max_length=200),
                                            'category': sanitize_text(category, max_length=200) if category else None,
                                            'year': year
                                        })

                            if awards:
                                next_data['literary_awards'] = awards

                    break  # Only process first Work object

        except (json.JSONDecodeError, KeyError, AttributeError) as e:
            # If JSON parsing fails, return empty dict
            pass

    return next_data


def extract_from_schema(soup: BeautifulSoup) -> dict[str, Any]:
    """Extract book metadata from JSON-LD schema.

    Args:
        soup: BeautifulSoup parsed HTML

    Returns:
        Dictionary with book metadata from schema
    """
    schema_data = {}

    # Find JSON-LD script tag
    script_tag = soup.find('script', type='application/ld+json')
    if script_tag and script_tag.string:
        try:
            data = json.loads(script_tag.string)

            # Extract ISBN (may be ISBN-10 or ISBN-13)
            if 'isbn' in data:
                isbn_str = str(data['isbn'])
                validated_isbn = validate_isbn(isbn_str)
                if validated_isbn:
                    if len(validated_isbn) == 13:
                        schema_data['isbn13'] = validated_isbn
                        # Try to derive ISBN-10 if it's ISBN-13
                        if validated_isbn.startswith('978'):
                            # Can't reliably convert back to ISBN-10, leave as None
                            pass
                    elif len(validated_isbn) == 10:
                        schema_data['isbn'] = validated_isbn

            # Extract page count
            if 'numberOfPages' in data:
                try:
                    schema_data['page_count'] = int(data['numberOfPages'])
                except (ValueError, TypeError):
                    pass

            # Extract language
            if 'inLanguage' in data:
                schema_data['language'] = sanitize_text(data['inLanguage'])

            # Extract cover image
            if 'image' in data:
                schema_data['cover_image_url'] = data['image']

            # Extract rating
            if 'aggregateRating' in data and isinstance(data['aggregateRating'], dict):
                rating_info = data['aggregateRating']
                if 'ratingValue' in rating_info:
                    try:
                        schema_data['average_rating'] = float(rating_info['ratingValue'])
                    except (ValueError, TypeError):
                        pass
                if 'ratingCount' in rating_info:
                    try:
                        schema_data['ratings_count'] = int(rating_info['ratingCount'])
                    except (ValueError, TypeError):
                        pass

            # Publisher is sometimes in the schema (though not common on Goodreads)
            if 'publisher' in data:
                if isinstance(data['publisher'], dict):
                    schema_data['publisher'] = sanitize_text(data['publisher'].get('name', ''))
                else:
                    schema_data['publisher'] = sanitize_text(str(data['publisher']))

        except (json.JSONDecodeError, KeyError, AttributeError) as e:
            # If JSON parsing fails, return empty dict (will fall back to HTML parsing)
            pass

    return schema_data


def extract_isbn(soup: BeautifulSoup) -> dict[str, Any]:
    """Extract ISBN and ISBN-13 from book page.

    Args:
        soup: BeautifulSoup parsed HTML

    Returns:
        Dictionary with 'isbn' and 'isbn13' keys
    """
    isbn_data = {"isbn": None, "isbn13": None}

    # Look for ISBN in book details section
    # Common pattern: <span>ISBN</span> <div>number</div>
    isbn_label = soup.find(string=lambda s: s and 'isbn' in s.lower())
    if isbn_label:
        parent = isbn_label.find_parent()
        if parent:
            # Try to find ISBN value in next sibling or nearby element
            isbn_value = parent.find_next_sibling()
            if isbn_value:
                isbn_text = sanitize_text(isbn_value.get_text())
                if isbn_text:
                    validated_isbn = validate_isbn(isbn_text)
                    if validated_isbn:
                        if len(validated_isbn) == 13:
                            isbn_data['isbn13'] = validated_isbn
                        else:
                            isbn_data['isbn'] = validated_isbn

    return isbn_data


def extract_publication_details(soup: BeautifulSoup) -> dict[str, Any]:
    """Extract publication year, publisher, page count, language.

    Args:
        soup: BeautifulSoup parsed HTML

    Returns:
        Dictionary with publication details
    """
    import re

    pub_data = {
        "publication_year": None,
        "publisher": None,
        "page_count": None,
        "language": None
    }

    # Look for publication info with data-testid
    pub_info = soup.find('p', {'data-testid': 'publicationInfo'})
    if pub_info:
        text = pub_info.get_text()

        # Extract year (4 digits)
        year_match = re.search(r'\b(1[0-9]{3}|20[0-9]{2})\b', text)
        if year_match:
            try:
                year = int(year_match.group(1))
                pub_data['publication_year'] = validate_publication_year(year)
            except (ValueError, TypeError):
                pass

        # Extract publisher (text after "by")
        if ' by ' in text:
            publisher = text.split(' by ')[-1].strip()
            pub_data['publisher'] = sanitize_text(publisher, max_length=200)

    # Note: Publisher information is not consistently available on Goodreads book pages.
    # The public book view pages do not include publisher data in the HTML or JSON-LD schema.
    # Publisher data may only be available on edition-specific pages or to logged-in users.
    # As a result, publisher will remain None for most books scraped from public pages.

    # Extract page count from pagesFormat
    pages_elem = soup.find('p', {'data-testid': 'pagesFormat'})
    if pages_elem:
        text = pages_elem.get_text()
        pages_match = re.search(r'(\d+)\s*pages', text, re.IGNORECASE)
        if pages_match:
            try:
                pub_data['page_count'] = int(pages_match.group(1))
            except ValueError:
                pass

    return pub_data


def extract_genres(soup: BeautifulSoup) -> list[str]:
    """Extract genre tags/shelves from book page.

    Args:
        soup: BeautifulSoup parsed HTML

    Returns:
        List of genre strings
    """
    genres = []

    # Look for genre buttons in the new Goodreads layout
    genre_elements = soup.find_all('span', class_='BookPageMetadataSection__genreButton')

    for element in genre_elements[:20]:  # Limit to 20 genres
        genre_text = sanitize_text(element.get_text())
        if genre_text:
            genres.append(genre_text)

    # Deduplicate
    return list(dict.fromkeys(genres))


def extract_rating_details(soup: BeautifulSoup) -> dict[str, Any]:
    """Extract average rating and ratings count.

    Args:
        soup: BeautifulSoup parsed HTML

    Returns:
        Dictionary with rating details
    """
    import re

    rating_data = {
        "average_rating": None,
        "ratings_count": None
    }

    # Look for rating in RatingStatistics div
    rating_div = soup.find('div', class_='RatingStatistics__rating')
    if rating_div:
        try:
            rating_text = rating_div.get_text(strip=True)
            rating_data['average_rating'] = float(rating_text)
        except (ValueError, AttributeError):
            pass

    # Ratings count in RatingStatistics__meta
    meta_div = soup.find('div', class_='RatingStatistics__meta')
    if meta_div:
        text = meta_div.get_text()
        # Extract numbers like "123,456 ratings"
        count_match = re.search(r'([\d,]+)\s+ratings', text)
        if count_match:
            try:
                count_text = count_match.group(1).replace(',', '')
                rating_data['ratings_count'] = int(count_text)
            except (ValueError, AttributeError):
                pass

    return rating_data


def extract_cover_image(soup: BeautifulSoup) -> str | None:
    """Extract book cover image URL.

    Args:
        soup: BeautifulSoup parsed HTML

    Returns:
        Cover image URL or None
    """
    # Look for ResponsiveImage class (book cover)
    img = soup.find('img', class_='ResponsiveImage')
    if img and img.get('src'):
        return img.get('src')

    return None

File: parser-ts/src/parsers/book-parser.ts
import * as cheerio from 'cheerio';
import { Book, LiteraryAward } from '../models/book.model';
import { DataValidator } from '../validators/data-validator';

export class BookParser {
  /**
   * Parse a Goodreads book detail page
   * Uses __NEXT_DATA__ as primary source (like Python implementation), with HTML fallbacks
   */
  static parseBookPage(html: string, goodreadsUrl: string): Partial<Book> {
    const $ = cheerio.load(html);

    // Extract from __NEXT_DATA__ first (most reliable source)
    const nextData = this.extractFromNextData($);

    const goodreadsId = this.extractGoodreadsId(goodreadsUrl);
    const title = nextData.title || this.extractTitle($);
    const author = nextData.author || this.extractAuthor($);
    const additionalAuthors = this.extractAdditionalAuthors($);
    const isbn = nextData.isbn || this.extractIsbn($);
    const isbn13 = nextData.isbn13 || this.extractIsbn13($);
    const publicationDate = nextData.publicationDate || this.extractPublicationDate($);
    const publisher = nextData.publisher || this.extractPublisher($);
    const pageCount = nextData.pageCount || this.extractPageCount($);
    const language = nextData.language || this.extractLanguage($);
    const setting = nextData.setting || this.extractSetting($);
    const literaryAwards = nextData.literaryAwards.length > 0 ? nextData.literaryAwards : this.extractLiteraryAwards($);
    const genres = this.extractGenres($);
    const averageRating = nextData.averageRating || this.extractAverageRating($);
    const ratingsCount = nextData.ratingsCount || this.extractRatingsCount($);
    const coverImageUrl = nextData.coverImageUrl || this.extractCoverImageUrl($);

    return {
      goodreadsId,
      title,
      author,
      additionalAuthors,
      isbn,
      isbn13,
      publicationDate,
      publisher,
      pageCount,
      language,
      setting,
      literaryAwards,
      genres: DataValidator.normalizeGenres(genres),
      averageRating,
      ratingsCount,
      coverImageUrl,
      goodreadsUrl,
    };
  }

  /**
   * Extract book metadata from __NEXT_DATA__ Apollo state
   * This is the primary data source - matches Python implementation
   */
  private static extractFromNextData($: cheerio.CheerioAPI): {
    title: string | null;
    author: string | null;
    isbn: string | null;
    isbn13: string | null;
    publicationDate: string | null;
    publisher: string | null;
    pageCount: number | null;
    language: string | null;
    setting: string | null;
    literaryAwards: LiteraryAward[];
    averageRating: number | null;
    ratingsCount: number | null;
    coverImageUrl: string | null;
  } {
    const result: any = {
      title: null,
      author: null,
      isbn: null,
      isbn13: null,
      publicationDate: null,
      publisher: null,
      pageCount: null,
      language: null,
      setting: null,
      literaryAwards: [],
      averageRating: null,
      ratingsCount: null,
      coverImageUrl: null,
    };

    try {
      const scriptTag = $('script#__NEXT_DATA__').first();
      if (!scriptTag.length) return result;

      const jsonText = scriptTag.html();
      if (!jsonText) return result;

      const data = JSON.parse(jsonText);
      const apolloState = data?.props?.pageProps?.apolloState;
      if (!apolloState) return result;

      // Find Book object (key starts with "Book:")
      for (const [key, value] of Object.entries(apolloState)) {
        if (key.startsWith('Book:') && typeof value === 'object' && value !== null) {
          const bookData = value as any;
          const details = bookData.details;

          if (details) {
            // Extract ISBN and ISBN-13
            if (details.isbn) {
              const isbnStr = String(details.isbn);
              if (isbnStr.length === 10) {
                result.isbn = isbnStr;
              }
            }
            if (details.isbn13) {
              const isbn13Str = String(details.isbn13);
              if (isbn13Str.length === 13) {
                result.isbn13 = isbn13Str;
              }
            }

            // Publisher
            if (details.publisher) {
              result.publisher = String(details.publisher).trim().substring(0, 200);
            }

            // Page count
            if (details.numPages) {
              result.pageCount = parseInt(details.numPages, 10);
            }

            // Language
            if (details.language?.name) {
              result.language = String(details.language.name).trim();
            }

            // Publication date from publicationTime (milliseconds timestamp)
            if (details.publicationTime) {
              try {
                const timestamp = parseInt(details.publicationTime, 10);
                const date = new Date(timestamp);
                result.publicationDate = date.toLocaleDateString('en-US', {
                  year: 'numeric',
                  month: 'long',
                  day: 'numeric',
                });
              } catch (e) {
                // Ignore date parsing errors
              }
            }
          }

          // Title
          if (bookData.title) {
            result.title = String(bookData.title).trim();
          }

          // Primary author
          if (bookData.primaryContributorEdge?.node?.name) {
            result.author = String(bookData.primaryContributorEdge.node.name).trim();
          }

          break; // Only process first Book object
        }
      }

      // Find Work object (key starts with "Work:") for setting and awards
      for (const [key, value] of Object.entries(apolloState)) {
        if (key.startsWith('Work:') && typeof value === 'object' && value !== null) {
          const workData = value as any;
          const details = workData.details;

          if (details) {
            // Setting - extract from places array
            if (details.places?.length > 0 && details.places[0]?.name) {
              result.setting = String(details.places[0].name).trim().substring(0, 200);
            }

            // Literary Awards - extract from awardsWon array
            if (details.awardsWon && Array.isArray(details.awardsWon)) {
              for (const award of details.awardsWon) {
                if (award?.name) {
                  const awardObj: any = {
                    name: String(award.name).trim().substring(0, 200),
                  };

                  // Category
                  if (award.category) {
                    awardObj.category = String(award.category).trim().substring(0, 200);
                  }

                  // Year from awardedAt timestamp
                  if (award.awardedAt) {
                    try {
                      const timestamp = parseInt(award.awardedAt, 10);
                      const date = new Date(timestamp);
                      awardObj.year = date.getFullYear();
                    } catch (e) {
                      // Ignore year parsing errors
                    }
                  }

                  result.literaryAwards.push(new LiteraryAward(awardObj));
                }
              }
            }
          }

          break; // Only process first Work object
        }
      }
    } catch (e) {
      // If JSON parsing fails, return what we have (empty result)
    }

    return result;
  }

  private static extractGoodreadsId(url: string): string {
    const match = url.match(/\/book\/show\/([^/?]+)/);
    return match ? match[1] : '';
  }

  private static extractTitle($: cheerio.CheerioAPI): string {
    const selectors = [
      'h1[data-testid="bookTitle"]',
      'h1.Text__title1',
      'h1#bookTitle',
      '.BookPageTitleSection h1',
    ];

    for (const selector of selectors) {
      const text = $(selector).first().text().trim();
      if (text) return text;
    }

    return '';
  }

  private static extractAuthor($: cheerio.CheerioAPI): string {
    const selectors = [
      '.ContributorLinksList .ContributorLink__name',
      'a[data-testid="name"]',
      '.authorName span[itemprop="name"]',
      '#bookAuthors .authorName',
    ];

    for (const selector of selectors) {
      const text = $(selector).first().text().trim();
      if (text) return text;
    }

    return '';
  }

  private static extractAdditionalAuthors($: cheerio.CheerioAPI): string[] {
    const authors: string[] = [];

    $('.ContributorLinksList .ContributorLink__name').each((index, element) => {
      if (index > 0) {
        // Skip first author (main author)
        const name = $(element).text().trim();
        if (name) authors.push(name);
      }
    });

    return authors;
  }

  private static extractIsbn($: cheerio.CheerioAPI): string | null {
    const text = $('span[itemprop="isbn"]').first().text().trim();
    return text || null;
  }

  private static extractIsbn13($: cheerio.CheerioAPI): string | null {
    const selectors = [
      '[data-testid="isbn13"]',
      'div:contains("ISBN13") + div',
    ];

    for (const selector of selectors) {
      let text = $(selector).first().text().trim();
      // Remove "ISBN13:" prefix if present
      text = text.replace(/^ISBN13:\s*/i, '');
      if (text) return text;
    }

    return null;
  }

  private static extractPublicationDate($: cheerio.CheerioAPI): string | null {
    const selectors = [
      '[data-testid="publicationInfo"]',
      'div[class*="FeaturedDetails"] p:contains("Published")',
      'div.row:contains("Published")',
    ];

    for (const selector of selectors) {
      const text = $(selector).first().text().trim();
      if (text) {
        // Extract date from text like "Published January 1st 2000"
        const dateMatch = text.match(/([A-Z][a-z]+ \d{1,2}(?:st|nd|rd|th)?,? \d{4})/);
        return dateMatch ? dateMatch[1] : text;
      }
    }

    return null;
  }

  private static extractPublisher($: cheerio.CheerioAPI): string | null {
    const text = $('[data-testid="publisher"]').first().text().trim();
    return text || null;
  }

  private static extractPageCount($: cheerio.CheerioAPI): number | null {
    const selectors = [
      '[data-testid="pagesFormat"]',
      'span[itemprop="numberOfPages"]',
      'div.row:contains("pages")',
    ];

    for (const selector of selectors) {
      const text = $(selector).first().text().trim();
      const match = text.match(/(\d+)\s*pages/);
      if (match) {
        return parseInt(match[1], 10);
      }
    }

    return null;
  }

  private static extractLanguage($: cheerio.CheerioAPI): string | null {
    const text = $('[itemprop="inLanguage"]').first().text().trim();
    return text || null;
  }

  private static extractSetting($: cheerio.CheerioAPI): string | null {
    // Settings are less standardized, try to find them
    const settingText = $('.infoBoxRowItem:contains("Setting")').next().text().trim();
    return settingText || null;
  }

  private static extractLiteraryAwards($: cheerio.CheerioAPI): LiteraryAward[] {
    const awards: LiteraryAward[] = [];

    $('.awardsAndBadges .LiteraryAward, .infoBoxRowItem a[href*="/award/"]').each((_, element) => {
      const name = $(element).text().trim();
      if (name) {
        awards.push(new LiteraryAward({ name }));
      }
    });

    return awards;
  }

  private static extractGenres($: cheerio.CheerioAPI): string[] {
    const genres: string[] = [];

    // New Goodreads layout
    $('[data-testid="genresList"] a, .BookPageMetadataSection__genres a').each((_, element) => {
      const genre = $(element).text().trim();
      if (genre) genres.push(genre);
    });

    // Old layout
    if (genres.length === 0) {
      $('.elementList .left .bookPageGenreLink').each((_, element) => {
        const genre = $(element).text().trim();
        if (genre) genres.push(genre);
      });
    }

    return genres;
  }

  private static extractAverageRating($: cheerio.CheerioAPI): number | null {
    const selectors = [
      '[data-testid="averageRating"]',
      '.RatingStatistics__rating',
      'span[itemprop="ratingValue"]',
    ];

    for (const selector of selectors) {
      const text = $(selector).first().text().trim();
      const rating = parseFloat(text);
      if (!isNaN(rating)) return rating;
    }

    return null;
  }

  private static extractRatingsCount($: cheerio.CheerioAPI): number | null {
    const selectors = [
      '[data-testid="ratingsCount"]',
      'span[itemprop="ratingCount"]',
      'span[data-testid="ratingsCount"]',
    ];

    for (const selector of selectors) {
      const text = $(selector).first().text().trim();
      // Extract number from text like "123,456 ratings"
      const match = text.match(/([\d,]+)\s*rating/);
      if (match) {
        return parseInt(match[1].replace(/,/g, ''), 10);
      }
    }

    return null;
  }

  private static extractCoverImageUrl($: cheerio.CheerioAPI): string | null {
    const selectors = [
      'img[data-testid="coverImage"]',
      '.BookCover__image img',
      'img[id="coverImage"]',
      'img[class*="ResponsiveImage"]',
    ];

    for (const selector of selectors) {
      const src = $(selector).first().attr('src');
      if (src) {
        // Use highest quality version
        return src.replace(/_SX\d+_/, '_SX1200_').replace(/_SY\d+_/, '_SY1200_');
      }
    }

    return null;
  }
}
